debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/xstream/XstreamStreamingChangeEventSource.java

```diff
@@ -71,11 +76,26 @@ public void execute(ChangeEventSourceContext context) throws InterruptedExceptio
                     startPosition, 1, 1, XStreamOut.DEFAULT_MODE);
 
             LcrEventHandler handler = new LcrEventHandler(errorHandler, dispatcher, clock, schema, offsetContext, this.tablenameCaseInsensitive);
+            Metronome metronome = Metronome.sleeper(Duration.ofSeconds(1), Clock.SYSTEM);
 
             // 2. receive events while running
-            while (context.isRunning()) {
-                LOGGER.trace("Receiving LCR");
-                xsOut.receiveLCRCallback(handler, XStreamOut.DEFAULT_MODE);
+            while (true) {
+                LCR lcr = xsOut.receiveLCR(XStreamOut.DEFAULT_MODE);
+                if (xsOut.getBatchStatus() == XStreamOut.EXECUTING) {
+                    processLCR(lcr, handler);
+                }
+                else {
+                    // When a batch begins, Xstream expects that to end before a new operation can
+                    // be performed, which includes detachment. This is why we only test for context
+                    // shutdown once we've left the LCR batch execution status.
+                    if (!context.isRunning()) {
+                        break;
+                    }
+
+                    // Xstream callback API implies a 1 second delay between batches
+                    // To mimic that behavior with the non-callback API, we apply a pause of 1 second
+                    metronome.pause();
```

Comment:
So to be clear, is this something that happens "often"? Or only rarely? I.e. is there any impact on latency in the common case? Or would it have been the same with the previous approach, too?

Reply:
So the callback API, e.g. `receiveLCRCallback` internally applies a 1 second delay between batches, and goal was to keep that the same.  This meant that if an LCR was returned, we wouldn't apply the pause until we had drained all the LCR records from Xstreams and we would then apply a 1 second poll until a new record was detected.